pvanalytics/features/shading.py

```diff
@@ -0,0 +1,340 @@
+            ndimage.binary_hit_or_miss(wires, np.flipud(j), np.flipud(k))
+        )
+    )
+    return np.logical_and(a1, a2)
+
+
+def _fill_gaps(wires):
+    """Fill 1-minute gaps."""
+    mask = np.array([[1, 0, 1]])
+    return np.logical_or(wires, ndimage.binary_hit_or_miss(wires, mask))
+
+
+def _remove_spikes(wires):
+    """Remove 1-day spikes."""
+    mask = np.array([[0, 1, 0]]).T
+    return np.logical_and(
+        wires,
+        np.logical_not(ndimage.binary_hit_or_miss(wires, mask))
+    )
+
+
+def _restore_gaps(wires):
+    """Restore 2-minute and 1-minute gaps"""
+    wires = np.logical_or(
+        wires,
+        ndimage.binary_hit_or_miss(
+            wires,
+            np.array([[0, 1, 0, 0, 1]]), np.array([[0, 0, 1, 1, 0]])
+        )
+    )
+    return np.logical_or(
+        wires,
+        ndimage.binary_hit_or_miss(wires, np.array([[1, 0, 1]]))
+    )
+
+
+def _width(bbox):
+    """Return the width of a bounding box.
+
+    Parameters
+    ----------
+    bbox : tuple
+        4-tuple specifying ``(min_row, min_col, max_row, max_col)``.
+
+    Returns
+    -------
+    int
+        The width of the bounding box in pixels.
+    """
+    return bbox[3] - bbox[1] + 1
+
+
+def _height(bbox):
+    """Return the height of a bounding box.
+
+    Parameters
+    ----------
+    bbox : tuple
+        4-tuple specifying ``(min_row, min_col, max_row, max_col)``.
+
+    Returns
+    -------
+    int
+        The height of the bounding box in pixels.
+    """
+    return bbox[2] - bbox[0] + 1
+
+
+def _filter_blobs(wires, blob_length, connectivity):
+    """Remove blobs shorter than blob_length.
+
+    Examines all connected components in the image and removes any
+    that span fewer than `blob_length` columns.
+    """
+    components = measure.label(wires, connectivity=connectivity, background=0)
+    keep = [props.label for props in measure.regionprops(components)
+            if _height(props.bbox) > blob_length]
+    return np.isin(components, keep)
+
+
+def _tand(degrees):
+    """Tangent of an angle in degrees."""
+    return np.tan(np.deg2rad(degrees))
+
+
+def _filter_bars(wires, out):
+    """Only keep shadows that are 'bar-shaped' and span multiple days."""
+    for angle in range(0, 90, 5):
+        if angle < 75:
+            dim = 20
+            mid = 10
+        else:
+            dim = round(10 + 2 * angle // 5)
+            mid = round(6 + angle // 5)
+        cc = np.arange(0, dim)
+        se = np.zeros((dim, dim))
+        if angle <= 45:
+            rr = mid + np.round((cc - mid) * _tand(angle))
+            se[cc, rr.astype('int')] = 1
+        else:
+            rr = mid + np.round((cc - mid) * _tand(90 - angle))
+            se[cc, rr.astype('int')] = 1
+            se = se.T
+        out = np.logical_or(
+            out,
+            morphology.binary_opening(wires, se)
+        )
+        out = np.logical_or(
+            out,
+            morphology.binary_opening(wires, np.flipud(se))
+        )
```

Comment:
These two operations are creating 4 copies of the image, we can reduce this by allocating one `ndarray` that `morphology.binary_opening` puts its output into and passing `out=out` to `np.logical_or`. Similar performance improvements can be had in the other helper functions used bye `_clean_wires`.